### PR TITLE
Refactor - FatalMandatory Function

### PR DESCRIPTION
**Pull request by RossMurr4y**
## Description

* Refactor the `fatalMandatory()` bash function, providing specific errors as to which variable was required but missing. 
* set a tidy value for the `PS4` environment variable, allowing bash to output an erroring calling functions features

#### Usage
```bash
# return a status
$ return_on_invalid_environment_variables 255 "ENV_VAR" "ENV_VAR2"
(Fatal)Mandatory argument missing: ENV_VAR. Check usage via -h option.
(Fatal)Mandatory argument missing: ENV_VAR2. Check usage via -h option.

# error immediately
exit_on_invalid_environment_variables "GENERATION_DIR" "NOT_REAL"
/home/hamlet/cmdb/hamlet/executor/cli
bash: !i: Undefined mandatory environment variable: NOT_REAL
```

## Motivation and Context

Current error messages for fatalMandatory provide little advice as to what variable is missing. This is compounded by nested bash scripts using the argument.

## How Has This Been Tested?

Local testing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions

- [x] None

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] None of the above.

### CODE REVIEW

**Comment by RossMurr4y** [comment]
I wonder now if there should be a 3rd option to `exit` or `return` - `compile_list` - which would add the missing variable to a growing global variable. At the end of the script (primarily I am thinking of `setContext.sh` here) we could then error or return based on the list length and provide a full list of missing variables.

Thoughts?

**Comment by ml019** [comment]
Keep it simple. If it takes a couple of runs to resolve any missing variables, I don't see that as an issue.

File: execution/manageJSON.sh [hunk 1]
```diff
@@ -77,8 +77,8 @@ JSON_ARRAY=(${JSON_LIST})
 JSON_ARRAY+=("$@")
 
 # Ensure mandatory arguments have been provided
-[[ (-z "${JSON_OUTPUT}") ||
-    ("${#JSON_ARRAY[@]}" -eq 0) ]] && fatalMandatory && exit 255
+validate_environment_variables "exit" 255 "JSON_OUTPUT"
+[[ ("${#JSON_ARRAY[@]}" -eq 0) ]] && fatalMandatory && exit 255
```
Comment (ml019) [inline execution/manageJSON.sh:81]: Change to fatalMandatory means this use needs adjustment

File: execution/utility.sh [hunk 2]
```diff
@@ -2518,3 +2515,32 @@ function get_openapi_definition_filename() {
...
+function validate_environment_variables(){
+  local on_fail="$1"; shift
+  local status="$1"; shift
+  local parts=("$@")
+
+  for i in "${parts[@]}"; do
+    case "${on_fail}" in
+      exit)
+        # Test the Env Var exists, and terminate immediately if its missing.
+        printf "%s\n" "${!i:?Undefined mandatory environment variable: ${i}}"
```
Comment (ml019) [inline execution/utility.sh:2528]: Won't this still produce an extra line space?

File: execution/utility.sh [hunk 3]
```diff
@@ -2518,3 +2515,32 @@ function get_openapi_definition_filename() {
   echo -n "defn-${name}-${accountId}-${region}-definition.json"
 }
 
+# -- Evaluate Mandatory Environment Variables --
+function validate_environment_variables(){
+  local on_fail="$1"; shift
+  local status="$1"; shift
+  local parts=("$@")
+
+  for i in "${parts[@]}"; do
```
Comment (ml019) [inline execution/utility.sh:2524]: How is i being tested to see if it is valid?

File: execution/utility.sh [hunk 4]
```diff
@@ -151,7 +155,9 @@ function fatalDirectory() {
 }
 
 function fatalMandatory() {
-  fatal "Mandatory arguments missing. Check usage via -h option."
+  local name="$1"; shift
+  local status="$1"; shift
+  fatal "Mandatory argument missing: ${name}. Check usage via -h option." && return $status
```
Comment (ml019) [inline execution/utility.sh:160]: Won't this just return from this function whereas the use requires returning out of the caller function.

**Review by ml019 — CHANGES_REQUESTED** [review]

**Review by RossMurr4y — COMMENTED** [review]
@ml019 I have updated the PR to reflect your comments. Would you mind taking another look?

File: execution/manageJSON.sh [hunk 5]
```diff
@@ -77,8 +77,8 @@ JSON_ARRAY=(${JSON_LIST})
 JSON_ARRAY+=("$@")
 
 # Ensure mandatory arguments have been provided
-[[ (-z "${JSON_OUTPUT}") ||
-    ("${#JSON_ARRAY[@]}" -eq 0) ]] && fatalMandatory && exit 255
+validate_environment_variables "exit" 255 "JSON_OUTPUT"
```
Comment (ml019) [inline execution/manageJSON.sh:80]: `255` needed? Need to check the return status?

**Review by ml019 — CHANGES_REQUESTED** [review]
I think a general consistency check across the new functions is needed. I think the signatures may have evolved but not all usage has kept up.

**Comment by ml019** [comment]
@RossMurr4y back to you

File: execution/utility.sh [hunk 6]
```diff
@@ -2518,3 +2514,38 @@ function get_openapi_definition_filename() {
...
+}
+
+function exit_on_invalid_environment_variables(){
+  local parts=("$@")
+  validate_environment_variables true "${parts[@]}"
+}
+
+function check_for_invalid_environment_variables(){
+  local parts=("$@")
+  validate_environment_variables "${parts[@]}"
```
[the remark this answers is not in the document]
Comment (RossMurr4y) [reply execution/utility.sh:2550]: No, there's a check to see if the first value is `true` but no action on false as its redundant. If the first arg passed is not true, it's assumed to just be another environment variable to check for.

File: automation/jenkins/aws/manageRepo.sh [hunk 7]
```diff
@@ -68,11 +68,11 @@ function init() {
         git init .
     fi
 
-    [[ (-z "${REPO_REMOTE}") ]] && fatalMandatory && return 1
+    return_on_invalid_environment_variables "REPO_REMOTE" || return $?
 
     git remote show "${REPO_REMOTE}" >/dev/null 2>&1
     if [[ $? -ne 0 ]]; then
-        [[ -z "${REPO_URL}" ]] && fatalMandatory && return 1
+        return_on_invalid_environment_variables 1 "REPO_URL" || return $?
```
[the remark this answers is not in the document]
Comment (RossMurr4y) [reply automation/jenkins/aws/manageRepo.sh]: Yep, looks like I missed a couple uses of the `return_on_invalid_environment_variables` function when I renamed the function name. I've tidied it up now.

File: execution/manageJSON.sh (hunk 5 above)
Comment (RossMurr4y) [reply execution/manageJSON.sh]: Missed this one too, have updated it to use `exit_on_invalid_environment_variables` instead

**Comment by RossMurr4y** [comment]
@ml019 I believe I've caught them all now. Hopefully ...

**Review by ml019 — APPROVED** [review]
Nice and neat now.

**Pull request merged**